File: event_guests.go
// event_guests.go
package main

import (
	"encoding/csv"
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"time"
	"github.com/google/uuid"
)

type Guest struct {
	ID       string `json:"id"`
	Name     string `json:"name"`
	Email    string `json:"email"`
	RSVP     string `json:"rsvp"`
	PlusOne  bool   `json:"plus_one"`
	Dietary  string `json:"dietary"`
	Notes    string `json:"notes"`
	Added    string `json:"added"`
}

type Event struct {
	ID      string  `json:"id"`
	Name    string  `json:"name"`
	Date    string  `json:"date"`
	Venue   string  `json:"venue"`
	Guests  []Guest `json:"guests"`
	Created string  `json:"created"`
}

type Organizer struct {
	Events []Event `json:"events"`
}

var dataFile = "events.json"

func (o *Organizer) load() {
	data, err := os.ReadFile(dataFile)
	if err != nil {
		return
	}
	json.Unmarshal(data, o)
}

func (o *Organizer) save() {
	data, _ := json.MarshalIndent(o, "", "  ")
	os.WriteFile(dataFile, data, 0644)
}

func (o *Organizer) getEvent(id string) *Event {
	for i := range o.Events {
		if o.Events[i].ID == id {
			return &o.Events[i]
		}
	}
	return nil
}

func (o *Organizer) getGuest(e *Event, id string) *Guest {
	for i := range e.Guests {
		if e.Guests[i].ID == id {
			return &e.Guests[i]
		}
	}
	return nil
}

func (o *Organizer) create(name, date, venue string) {
	e := Event{
		ID:      uuid.New().String()[:8],
		Name:    name,
		Date:    date,
		Venue:   venue,
		Created: time.Now().Format(time.RFC3339),
	}
	o.Events = append(o.Events, e)
	o.save()
	fmt.Printf("✅ Event created: %s (ID: %s)\n", e.Name, e.ID)
}

func (o *Organizer) addGuest(eventID, name, email string, plusOne bool, dietary, notes string) {
	e := o.getEvent(eventID)
	if e == nil {
		fmt.Printf("Event %s not found.\n", eventID)
		return
	}
	g := Guest{
		ID:      uuid.New().String()[:8],
		Name:    name,
		Email:   email,
		RSVP:    "Pending",
		PlusOne: plusOne,
		Dietary: dietary,
		Notes:   notes,
		Added:   time.Now().Format(time.RFC3339),
	}
	e.Guests = append(e.Guests, g)
	o.save()
	fmt.Printf("✅ Guest added: %s (ID: %s)\n", g.Name, g.ID)
}

func (o *Organizer) listGuests(eventID, status string) {
	e := o.getEvent(eventID)
	if e == nil {
		fmt.Printf("Event %s not found.\n", eventID)
		return
	}
	guests := e.Guests
	if status != "" {
		var filtered []Guest
		for _, g := range guests {
			if g.RSVP == status {
				filtered = append(filtered, g)
			}
		}
		guests = filtered
	}
	if len(guests) == 0 {
		fmt.Println("No guests.")
		return
	}
	fmt.Printf("\n📋 Event: %s\n", e.Name)
	if e.Date != "" {
		fmt.Printf("Date: %s\n", e.Date)
	}
	if e.Venue != "" {
		fmt.Printf("Venue: %s\n", e.Venue)
	}
	fmt.Printf("\n👤 Guests (%d):\n", len(guests))
	for i, g := range guests {
		plus := " (plus‑one: Yes)"
		if !g.PlusOne {
			plus = " (plus‑one: No)"
		}
		dietary := ""
		if g.Dietary != "" {
			dietary = fmt.Sprintf(" 🥗 %s", g.Dietary)
		}
		notes := ""
		if g.Notes != "" {
			notes = fmt.Sprintf(" 📝 %s", g.Notes)
		}
		fmt.Printf("  %d. %s (%s) – %s%s%s%s\n", i+1, g.Name, g.Email, g.RSVP, plus, dietary, notes)
	}
}

func (o *Organizer) rsvp(eventID, guestID, status string) {
	e := o.getEvent(eventID)
	if e == nil {
		fmt.Printf("Event %s not found.\n", eventID)
		return
	}
	g := o.getGuest(e, guestID)
	if g == nil {
		fmt.Printf("Guest %s not found.\n", guestID)
		return
	}
	validStatuses := map[string]bool{"Pending": true, "Attending": true, "Declined": true}
	if !validStatuses[status] {
		fmt.Println("Invalid status. Choose: Pending, Attending, Declined")
		return
	}
	g.RSVP = status
	o.save()
	fmt.Printf("✅ %s RSVP updated to: %s\n", g.Name, status)
}

func (o *Organizer) stats(eventID string) {
	e := o.getEvent(eventID)
	if e == nil {
		fmt.Printf("Event %s not found.\n", eventID)
		return
	}
	total := len(e.Guests)
	attending, pending, declined, plusOnes := 0, 0, 0, 0
	for _, g := range e.Guests {
		switch g.RSVP {
		case "Attending":
			attending++
		case "Pending":
			pending++
		case "Declined":
			declined++
		}
		if g.PlusOne {
			plusOnes++
		}
	}
	fmt.Printf("\n📊 Event: %s\n", e.Name)
	fmt.Printf("  Total guests: %d\n", total)
	fmt.Printf("  Attending: %d\n", attending)
	fmt.Printf("  Pending: %d\n", pending)
	fmt.Printf("  Declined: %d\n", declined)
	fmt.Printf("  Plus‑ones: %d\n", plusOnes)
}

func (o *Organizer) search(eventID, term string) {
	e := o.getEvent(eventID)
	if e == nil {
		fmt.Printf("Event %s not found.\n", eventID)
		return
	}
	var results []Guest
	for _, g := range e.Guests {
		if contains(g.Name, term) || contains(g.Email, term) {
			results = append(results, g)
		}
	}
	if len(results) == 0 {
		fmt.Println("No matches.")
		return
	}
	fmt.Printf("\n🔍 Found %d guest(s):\n", len(results))
	for i, g := range results {
		fmt.Printf("  %d. %s (%s) – %s\n", i+1, g.Name, g.Email, g.RSVP)
	}
}

func contains(s, substr string) bool {
	return len(s) >= len(substr) && (s == substr || len(substr) == 0 ||
		(s[0:len(substr)] == substr) || (len(s) > len(substr) && (s[len(s)-len(substr):] == substr || 
		indexOfSubstr(s, substr) != -1)))
}

func indexOfSubstr(s, substr string) int {
	for i := 0; i <= len(s)-len(substr); i++ {
		if s[i:i+len(substr)] == substr {
			return i
		}
	}
	return -1
}

func (o *Organizer) export(eventID, filename string) {
	e := o.getEvent(eventID)
	if e == nil {
		fmt.Printf("Event %s not found.\n", eventID)
		return
	}
	f, err := os.Create(filename)
	if err != nil {
		fmt.Printf("Error creating file: %v\n", err)
		return
	}
	defer f.Close()
	w := csv.NewWriter(f)
	defer w.Flush()
	w.Write([]string{"ID", "Name", "Email", "RSVP", "Plus-One", "Dietary", "Notes", "Added"})
	for _, g := range e.Guests {
		w.Write([]string{g.ID, g.Name, g.Email, g.RSVP, fmt.Sprintf("%v", g.PlusOne), g.Dietary, g.Notes, g.Added})
	}
	fmt.Printf("✅ Exported %d guests to %s\n", len(e.Guests), filename)
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Usage: event_guests <command> [options]")
		return
	}
	o := &Organizer{}
	o.load()
	cmd := os.Args[1]

	switch cmd {
	case "create":
		createCmd := flag.NewFlagSet("create", flag.ExitOnError)
		name := createCmd.String("name", "", "")
		date := createCmd.String("date", "", "")
		venue := createCmd.String("venue", "", "")
		createCmd.Parse(os.Args[2:])
		if *name == "" && len(createCmd.Args()) > 0 {
			*name = createCmd.Args()[0]
		}
		if *name == "" {
			fmt.Println("create requires a name")
			return
		}
		o.create(*name, *date, *venue)

	case "add-guest":
		addCmd := flag.NewFlagSet("add-guest", flag.ExitOnError)
		eventID := addCmd.String("event-id", "", "")
		name := addCmd.String("name", "", "")
		email := addCmd.String("email", "", "")
		plusOne := addCmd.Bool("plus-one", false, "")
		dietary := addCmd.String("dietary", "", "")
		notes := addCmd.String("notes", "", "")
		addCmd.Parse(os.Args[2:])
		args := addCmd.Args()
		if *eventID == "" && len(args) > 0 {
			*eventID = args[0]
		}
		if *name == "" && len(args) > 1 {
			*name = args[1]
		}
		if *email == "" && len(args) > 2 {
			*email = args[2]
		}
		if *eventID == "" || *name == "" || *email == "" {
			fmt.Println("add-guest requires event-id, name, email")
			return
		}
		o.addGuest(*eventID, *name, *email, *plusOne, *dietary, *notes)

	case "list-guests":
		listCmd := flag.NewFlagSet("list-guests", flag.ExitOnError)
		eventID := listCmd.String("event-id", "", "")
		status := listCmd.String("status", "", "")
		listCmd.Parse(os.Args[2:])
		if *eventID == "" && len(listCmd.Args()) > 0 {
			*eventID = listCmd.Args()[0]
		}
		if *eventID == "" {
			fmt.Println("list-guests requires event-id")
			return
		}
		o.listGuests(*eventID, *status)

	case "rsvp":
		rsvpCmd := flag.NewFlagSet("rsvp", flag.ExitOnError)
		eventID := rsvpCmd.String("event-id", "", "")
		guestID := rsvpCmd.String("guest-id", "", "")
		status := rsvpCmd.String("status", "", "")
		rsvpCmd.Parse(os.Args[2:])
		if *eventID == "" && len(rsvpCmd.Args()) > 0 {
			*eventID = rsvpCmd.Args()[0]
		}
		if *guestID == "" && len(rsvpCmd.Args()) > 1 {
			*guestID = rsvpCmd.Args()[1]
		}
		if *status == "" && len(rsvpCmd.Args()) > 2 {
			*status = rsvpCmd.Args()[2]
		}
		if *eventID == "" || *guestID == "" || *status == "" {
			fmt.Println("rsvp requires event-id, guest-id, status")
			return
		}
		o.rsvp(*eventID, *guestID, *status)

	case "stats":
		statsCmd := flag.NewFlagSet("stats", flag.ExitOnError)
		eventID := statsCmd.String("event-id", "", "")
		statsCmd.Parse(os.Args[2:])
		if *eventID == "" && len(statsCmd.Args()) > 0 {
			*eventID = statsCmd.Args()[0]
		}
		if *eventID == "" {
			fmt.Println("stats requires event-id")
			return
		}
		o.stats(*eventID)

	case "search":
		searchCmd := flag.NewFlagSet("search", flag.ExitOnError)
		eventID := searchCmd.String("event-id", "", "")
		term := searchCmd.String("term", "", "")
		searchCmd.Parse(os.Args[2:])
		if *eventID == "" && len(searchCmd.Args()) > 0 {
			*eventID = searchCmd.Args()[0]
		}
		if *term == "" && len(searchCmd.Args()) > 1 {
			*term = searchCmd.Args()[1]
		}
		if *eventID == "" || *term == "" {
			fmt.Println("search requires event-id and term")
			return
		}
		o.search(*eventID, *term)

	case "export":
		exportCmd := flag.NewFlagSet("export", flag.ExitOnError)
		eventID := exportCmd.String("event-id", "", "")
		filename := exportCmd.String("filename", "guests.csv", "")
		exportCmd.Parse(os.Args[2:])
		if *eventID == "" && len(exportCmd.Args()) > 0 {
			*eventID = exportCmd.Args()[0]
		}
		if *eventID == "" {
			fmt.Println("export requires event-id")
			return
		}
		o.export(*eventID, *filename)

	default:
		fmt.Println("Unknown command. Use create, add-guest, list-guests, rsvp, stats, search, export.")
	}
}

File: event_guests.py
# event_guests.py
import json
import os
import sys
import argparse
import uuid
from datetime import datetime

DATA_FILE = "events.json"

class Guest:
    def __init__(self, name, email, plus_one=False, dietary="", notes="", guest_id=None):
        self.id = guest_id or str(uuid.uuid4())[:8]
        self.name = name
        self.email = email
        self.rsvp = "Pending"  # Pending, Attending, Declined
        self.plus_one = plus_one
        self.dietary = dietary
        self.notes = notes
        self.added = datetime.now().isoformat()

    def to_dict(self):
        return {
            "id": self.id,
            "name": self.name,
            "email": self.email,
            "rsvp": self.rsvp,
            "plus_one": self.plus_one,
            "dietary": self.dietary,
            "notes": self.notes,
            "added": self.added
        }

    @classmethod
    def from_dict(cls, data):
        g = cls(data["name"], data["email"], data.get("plus_one", False),
                data.get("dietary", ""), data.get("notes", ""), data.get("id"))
        g.rsvp = data.get("rsvp", "Pending")
        g.added = data.get("added", datetime.now().isoformat())
        return g

class Event:
    def __init__(self, name, date="", venue="", event_id=None):
        self.id = event_id or str(uuid.uuid4())[:8]
        self.name = name
        self.date = date
        self.venue = venue
        self.guests = []
        self.created = datetime.now().isoformat()

    def to_dict(self):
        return {
            "id": self.id,
            "name": self.name,
            "date": self.date,
            "venue": self.venue,
            "guests": [g.to_dict() for g in self.guests],
            "created": self.created
        }

    @classmethod
    def from_dict(cls, data):
        e = cls(data["name"], data.get("date", ""), data.get("venue", ""), data.get("id"))
        e.guests = [Guest.from_dict(g) for g in data.get("guests", [])]
        e.created = data.get("created", datetime.now().isoformat())
        return e

class Organizer:
    def __init__(self):
        self.events = []
        self.load()

    def load(self):
        if os.path.exists(DATA_FILE):
            with open(DATA_FILE, "r") as f:
                data = json.load(f)
                self.events = [Event.from_dict(e) for e in data]

    def save(self):
        with open(DATA_FILE, "w") as f:
            json.dump([e.to_dict() for e in self.events], f, indent=2)

    def get_event(self, event_id):
        for e in self.events:
            if e.id == event_id:
                return e
        return None

    def get_guest(self, event, guest_id):
        for g in event.guests:
            if g.id == guest_id:
                return g
        return None

    def create(self, name, date="", venue=""):
        e = Event(name, date, venue)
        self.events.append(e)
        self.save()
        print(f"✅ Event created: {e.name} (ID: {e.id})")

    def add_guest(self, event_id, name, email, plus_one=False, dietary="", notes=""):
        e = self.get_event(event_id)
        if not e:
            print(f"Event {event_id} not found.")
            return
        g = Guest(name, email, plus_one, dietary, notes)
        e.guests.append(g)
        self.save()
        print(f"✅ Guest added: {g.name} (ID: {g.id})")

    def list_guests(self, event_id, status=None):
        e = self.get_event(event_id)
        if not e:
            print(f"Event {event_id} not found.")
            return
        guests = e.guests
        if status:
            guests = [g for g in guests if g.rsvp.lower() == status.lower()]
        if not guests:
            print("No guests.")
            return
        print(f"\n📋 Event: {e.name}")
        if e.date:
            print(f"Date: {e.date}")
        if e.venue:
            print(f"Venue: {e.venue}")
        print(f"\n👤 Guests ({len(guests)}):")
        for i, g in enumerate(guests, 1):
            plus = " (plus‑one: Yes)" if g.plus_one else " (plus‑one: No)"
            dietary = f" 🥗 {g.dietary}" if g.dietary else ""
            notes = f" 📝 {g.notes}" if g.notes else ""
            print(f"  {i}. {g.name} ({g.email}) – {g.rsvp}{plus}{dietary}{notes}")

    def rsvp(self, event_id, guest_id, status):
        e = self.get_event(event_id)
        if not e:
            print(f"Event {event_id} not found.")
            return
        g = self.get_guest(e, guest_id)
        if not g:
            print(f"Guest {guest_id} not found.")
            return
        valid_statuses = ["Pending", "Attending", "Declined"]
        if status not in valid_statuses:
            print(f"Invalid status. Choose: {', '.join(valid_statuses)}")
            return
        g.rsvp = status
        self.save()
        print(f"✅ {g.name} RSVP updated to: {status}")

    def stats(self, event_id):
        e = self.get_event(event_id)
        if not e:
            print(f"Event {event_id} not found.")
            return
        total = len(e.guests)
        attending = sum(1 for g in e.guests if g.rsvp == "Attending")
        pending = sum(1 for g in e.guests if g.rsvp == "Pending")
        declined = sum(1 for g in e.guests if g.rsvp == "Declined")
        plus_ones = sum(1 for g in e.guests if g.plus_one)
        print(f"\n📊 Event: {e.name}")
        print(f"  Total guests: {total}")
        print(f"  Attending: {attending}")
        print(f"  Pending: {pending}")
        print(f"  Declined: {declined}")
        print(f"  Plus‑ones: {plus_ones}")

    def search(self, event_id, term):
        e = self.get_event(event_id)
        if not e:
            print(f"Event {event_id} not found.")
            return
        term_lower = term.lower()
        results = [g for g in e.guests if term_lower in g.name.lower() or term_lower in g.email.lower()]
        if not results:
            print("No matches.")
            return
        print(f"\n🔍 Found {len(results)} guest(s):")
        for i, g in enumerate(results, 1):
            print(f"  {i}. {g.name} ({g.email}) – {g.rsvp}")

    def export(self, event_id, filename):
        import csv
        e = self.get_event(event_id)
        if not e:
            print(f"Event {event_id} not found.")
            return
        with open(filename, 'w', newline='') as f:
            writer = csv.writer(f)
            writer.writerow(["ID", "Name", "Email", "RSVP", "Plus-One", "Dietary", "Notes", "Added"])
            for g in e.guests:
                writer.writerow([g.id, g.name, g.email, g.rsvp, g.plus_one, g.dietary, g.notes, g.added])
        print(f"✅ Exported {len(e.guests)} guests to {filename}")

def main():
    parser = argparse.ArgumentParser(description="Event Organizer - Guest Manager")
    subparsers = parser.add_subparsers(dest="cmd", required=True)

    create_parser = subparsers.add_parser("create")
    create_parser.add_argument("name")
    create_parser.add_argument("--date", default="")
    create_parser.add_argument("--venue", default="")

    add_parser = subparsers.add_parser("add-guest")
    add_parser.add_argument("event_id")
    add_parser.add_argument("name")
    add_parser.add_argument("email")
    add_parser.add_argument("--plus-one", action="store_true")
    add_parser.add_argument("--dietary", default="")
    add_parser.add_argument("--notes", default="")

    list_parser = subparsers.add_parser("list-guests")
    list_parser.add_argument("event_id")
    list_parser.add_argument("--status", choices=["Pending", "Attending", "Declined"])

    rsvp_parser = subparsers.add_parser("rsvp")
    rsvp_parser.add_argument("event_id")
    rsvp_parser.add_argument("guest_id")
    rsvp_parser.add_argument("status", choices=["Pending", "Attending", "Declined"])

    stats_parser = subparsers.add_parser("stats")
    stats_parser.add_argument("event_id")

    search_parser = subparsers.add_parser("search")
    search_parser.add_argument("event_id")
    search_parser.add_argument("term")

    export_parser = subparsers.add_parser("export")
    export_parser.add_argument("event_id")
    export_parser.add_argument("--filename", default="guests.csv")

    args = parser.parse_args()
    app = Organizer()

    if args.cmd == "create":
        app.create(args.name, args.date, args.venue)
    elif args.cmd == "add-guest":
        app.add_guest(args.event_id, args.name, args.email, args.plus_one, args.dietary, args.notes)
    elif args.cmd == "list-guests":
        app.list_guests(args.event_id, args.status)
    elif args.cmd == "rsvp":
        app.rsvp(args.event_id, args.guest_id, args.status)
    elif args.cmd == "stats":
        app.stats(args.event_id)
    elif args.cmd == "search":
        app.search(args.event_id, args.term)
    elif args.cmd == "export":
        app.export(args.event_id, args.filename)

if __name__ == "__main__":
    main()
